circle-mlir/circle-mlir/lib/dialect/src/ConstFold.h

```diff
@@ -0,0 +1,263 @@
+                                                         rhs.getSplatValue<ElementValueT>()));
+    if (!element_result)
+      return {};
+
+    return DenseElementsAttr::get(type, element_result);
+  }
+
+  auto num_elements = type.getNumElements();
+
+  SmallVector<ElementValueT, 16> new_values;
+  new_values.reserve(num_elements);
+  const auto result_shape = type.getShape();
+  std::vector<int64_t> current_index(type.getRank(), 0);
+  // Create the new shape with ones padded to the left.
+  const std::vector<int64_t> lhs_new_shape =
+    GetPaddedShape(lhs.getType().getShape(), type.getRank());
+  const std::vector<int64_t> rhs_new_shape =
+    GetPaddedShape(rhs.getType().getShape(), type.getRank());
+
+  auto lhs_old_values = lhs.getValues<ElementValueT>();
+  auto rhs_old_values = rhs.getValues<ElementValueT>();
+
+  // Add each pair of the corresponding values in the dense elements
+  // attributes.
+  for (int64_t i = 0; i < num_elements; ++i)
+  {
+    // current_index represents the index
+    // in the N-dimension tensor. GetElementIndex returns
+    // the index in the flat representation of the original tensor
+    // to use.
+    const int64_t lhs_index = lhs_is_splat ? 0 : GetElementIndex(lhs_new_shape, current_index);
+    const int64_t rhs_index = rhs_is_splat ? 0 : GetElementIndex(rhs_new_shape, current_index);
+
+    new_values.push_back(
+      calculate(*(lhs_old_values.begin() + lhs_index), *(rhs_old_values.begin() + rhs_index)));
+    IncrementIndex(result_shape, &current_index);
+  }
+  return DenseElementsAttr::get(type, ArrayRef<ElementValueT>(new_values));
+}
+
+/// Performs const folding `calculate` with broadcast behavior on the two
+/// attributes `operand1` and `operand2` and returns the result if possible.
+/// This function assumes the two operands are verified to have value
+/// attributes of broadcastable types.
+template <class AttrElementT, class ElementValueT = typename AttrElementT::ValueType,
+          class CalculationT = llvm::function_ref<ElementValueT(ElementValueT, ElementValueT)>>
+Attribute ConstFoldBinaryOp(Type result_type, Attribute operand1, Attribute operand2,
+                            const CalculationT &calculate)
+{
+  if (operand1.dyn_cast_or_null<DenseElementsAttr>() &&
+      operand2.dyn_cast_or_null<DenseElementsAttr>())
+  {
+    return ConstFoldBinaryOpDenseDense<AttrElementT, ElementValueT>(
+      result_type, operand1.cast<DenseElementsAttr>(), operand2.cast<DenseElementsAttr>(),
+      calculate);
+  }
+
+  // TODO: support other attribute kinds
+
+  return {};
+}
+
+/// Performs const folding with broadcast behavior on the two attributes in
+/// `operands` and returns the result if possible.
+/// Depending on the given `resultType`, either `floatCalculate` or
+/// `intCalculate` is chosen to conduct the calculate.
+Attribute ConstFoldBinaryOp(Type result_type, ArrayRef<Attribute> operands,
```

Comment:
It would be nice to add `inline` keyword to prevent multiple definition errors.
```suggestion
inline Attribute ConstFoldBinaryOp(Type result_type, ArrayRef<Attribute> operands,
```

Reply:
ditto